streams/shared/src/main/scala/zio/stream/ZTransducer.scala

```diff
@@ -120,6 +120,37 @@ abstract class ZTransducer[-R, +E, -I, +O](val push: ZManaged[R, Nothing, Option
       }
     }
 
+  /**
+   * Like [[zip]], but widens input type to `Either[I, I1]`.
+   */
+  def choose[R1 <: R, E1 >: E, I1, O1](
+    that: ZTransducer[R1, E1, I1, O1]
+  ): ZTransducer[R1, E1, Either[I, I1], Either[O, O1]] =
+    ZTransducer {
+      for {
+        p1 <- self.push
+        p2 <- that.push
+      } yield { (in: Option[Chunk[Either[I, I1]]]) =>
+        val (lefts, rights) = in.fold[(Option[Chunk[I]], Option[Chunk[I1]])]((None, None)) { ch =>
+          if (ch.isEmpty) {
+            (Some(Chunk.empty), Some(Chunk.empty))
+          } else {
+            val leftBuilder: ChunkBuilder[I]   = ChunkBuilder.make()
+            val rightBuilder: ChunkBuilder[I1] = ChunkBuilder.make()
+            ch.foreach {
+              case Left(v)  => leftBuilder += v
+              case Right(v) => rightBuilder += v
+            }
+            (Some(leftBuilder.result()), Some(rightBuilder.result()))
```

Comment:
Given the input `Chunk(Left(1), Right(1), Left(1), Right(1))`, this will produce `Chunk(Left(1), Left(1), Right(1), Right(1))`. That is- it reorders the elements.

This implementation should create chunks of consecutive left/right elements from `in` and feed them in order to `p1` and `p2`.

Reply:
I'm not sure if this requirement could be formulated as a chunking-independent algebraic property. For me it seems that 2 transducers live in parallel timelines and we are could demand only internal output ordering.
Let me think about it a bit more.

Reply:
Seems to me this should hold at least:
```scala
stream.transduce(id choose id).runCollect == stream.runCollect
```
then probably this too:
```scala
stream.transduce(f choose g).transduce(h choose i).runCollect ==
  stream.transduce((f >>> h) choose (g >>> i)).runCollect
```

Reply:
@regiskuckaertz great findings!
The second law being rewritten as `(f choose g) >>> (h choose i) == (f >>> h) choose (g >>> i)` looks even cooler

Reply:
Oh yeah nice 😍